File: Septermber'21/add-two-numbers/add-two-numbers.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

// #define ll long long int

class Solution {
public:
    
    void append(ListNode** head_ref, int new_data) {
        ListNode* new_node = new ListNode();
        
        ListNode* last = *head_ref;
        new_node->val = new_data; 
        
        new_node->next = nullptr;
        
        if(*head_ref == NULL) {
            *head_ref = new_node;
            return;
        }
        
        while(last->next != NULL) {
            last = last->next;
        }
        last->next = new_node;
        return;
    }
    
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        
        int add;
        int flag = 0; 
        ListNode* l3 = NULL;
        while(l1 || l2) {
            add = 0;
            
            if(l1 != NULL){
                add += l1->val;
                l1 = l1->next;
            }
            
            if (l2 != NULL) {
                add += l2->val;
                l2 = l2->next;
            }

            
            add+=flag;
            flag = 0;
            
            if(add >=10) {
                add = add%10;
                flag = 1;
            }
            append(&l3, add);
            
            if(l1==NULL && l2==NULL && flag==1){
                append(&l3, flag);
            }
        }
        
        return l3;
    }
};

File: add-two-numbers/add-two-numbers.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        ListNode sum=new ListNode();
        add(0,l1,l2,sum);
        return sum.next;
    }
    public void add(int carry,ListNode l1,ListNode l2, ListNode sum){
        if(l1==null && l2==null){
            if(carry==1){
                sum.next=new ListNode(1);
                // sum.next.val=1;
                sum=sum.next;
            }
            return;
        }
        
        if(l1==null && l2!=null){
            while(l2!=null){
                sum.next=new ListNode();
                sum.next.val=(l2.val+carry)%10;
                sum=sum.next;
                if((l2.val+carry)>9)
                    carry=1;
                else
                    carry=0;
                l2=l2.next;
            }
            if(carry==1){
                sum.next=new ListNode(1);
                // sum.next.val=1;
                sum=sum.next;
            }
            return;
        }
        if(l1!=null && l2==null){
            while(l1!=null){
                sum.next=new ListNode();
                sum.next.val=(l1.val+carry)%10;
                sum=sum.next;
                if((l1.val+carry)>9)
                    carry=1;
                else
                    carry=0;
                l1=l1.next;
            }
            if(carry==1){
                sum.next=new ListNode(1);
                // sum.next.val=1;
                sum=sum.next;
            }
            return;
        }
        int s=l1.val+l2.val+carry;
        if(s>9)
            carry=1;
        else
            carry=0;
        sum.next=new ListNode();
        sum.next.val=s%10;
        sum=sum.next;
        add(carry,l1.next,l2.next,sum);
        return;
        
    }
}